### ISSUE TITLE
Inconsistency `CrossEntropyLoss` weights value between in gpu and in cpu.

### ISSUE THREAD

**Issue by GdoongMathew** [pytorch/pytorch#130193]
### 🐛 Describe the bug

There's an inconsistent results when the `CrossEntropyLoss` is placed in `cpu` and in `gpu`.
When the loss module is in `cpu`, if its weight value is updated by an in-place operation, the weight itself from the outside will also be updated.
```python
>>> from torch.nn import CrossEntropyLoss

>>> weight = torch.tensor([1., 2.])
... loss = CrossEntropyLoss(weight=weight)
... loss.weight *= 2
... assert torch.equal(weight, torch.tensor([1., 2.]))
AssertionError
```

But if the loss module is placed into a cuda device, then the weight value from the outside won't be updated.
```python
>>> weight = torch.tensor([1., 2.])
... loss = CrossEntropyLoss(weight=weight)
... loss.to("cuda")
... loss.weight *= 2
... assert torch.equal(weight, torch.tensor([1., 2.]))
```

This behavior might be related to #126854 .

### Versions

```
Collecting environment information...
PyTorch version: 2.2.2+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Home
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22631-SP0
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3050 4GB Laptop GPU
Nvidia driver version: 555.99
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:










Revision=

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] flake8-plugin-utils==1.3.3
[pip3] flake8-pytest-style==2.0.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.22.3
[pip3] onnx==1.11.0
[pip3] onnxruntime-training==1.15.1
[pip3] pytorch-lightning==2.1.3
[pip3] torch==2.2.2+cu121
[pip3] torch-directml==0.2.1.dev240521
[pip3] torch-ort==1.17.0
[pip3] torchaudio==2.2.1+cu121
[pip3] torchmetrics==1.3.2
[pip3] torchvision==0.17.2+cu121
[conda] Could not collect
```

**Comment by mikaylagawarecki** [comment]
I think this is expected behavior, `weight` is registered as a buffer here

https://github.com/pytorch/pytorch/blob/856fe230c708ed1aaba413b5e09f82bcb3b61019/torch/nn/modules/loss.py#L58

and in-place modification on buffer will modify it.

The reason why it doesn't happen for `.cuda()` is because `.cuda()` will change the TensorImpl associated (via swap_tensor or .data setting)

However, if you had initialized tensor on cuda before passing it in, you would observe the same behavior as in the first snippet

**Comment by GdoongMathew** [comment]
I understand that from the mechanism point of view, this might not seem like a bug, just wondering if this inconsistent behavior is anticipated.
